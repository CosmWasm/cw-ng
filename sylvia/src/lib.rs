#![cfg_attr(docsrs, feature(doc_cfg))]
//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here

pub mod builder;
pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
#[cfg_attr(docsrs, doc(cfg(feature = "cosmwasm_1_2")))]
#[cfg(feature = "cosmwasm_1_2")]
pub use cw_utils;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use derivative;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value, serde_json_wasm as serde_json,
};

pub mod replies {
    use cosmwasm_std::{DepsMut, Empty, Env, Event, MsgResponse};

    /// Represantation of `reply` context received in entry point.
    #[deprecated(
        since = "1.3.0",
        note = "This type is added temporarily to not break existing API. Since `2.0.0` it will replace the `sylvia::types::ReplyCtx` type."
    )]
    #[non_exhaustive]
    pub struct ReplyCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
        pub deps: DepsMut<'a, C>,
        pub env: Env,
        pub gas_used: u64,
        pub events: Vec<Event>,
        pub msg_responses: Vec<MsgResponse>,
    }

    #[allow(deprecated)]
    impl<'a, C: cosmwasm_std::CustomQuery>
        From<(DepsMut<'a, C>, Env, u64, Vec<Event>, Vec<MsgResponse>)> for ReplyCtx<'a, C>
    {
        fn from(
            (deps, env, gas_used, events, msg_responses): (
                DepsMut<'a, C>,
                Env,
                u64,
                Vec<Event>,
                Vec<MsgResponse>,
            ),
        ) -> Self {
            Self {
                deps,
                env,
                gas_used,
                events,
                msg_responses,
            }
        }
    }
}
