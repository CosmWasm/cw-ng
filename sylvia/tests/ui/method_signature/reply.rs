#![allow(unused_imports)]
use sylvia::cw_std::{Binary, Response, StdResult};
use sylvia::types::{InstantiateCtx, ReplyCtx};

pub mod mismatched_params {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=success)]
        fn first_reply(
            &self,
            _ctx: ReplyCtx,
            _data: Option<Binary>,
            param: String,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=failure)]
        fn second_reply(
            &self,
            _ctx: ReplyCtx,
            _data: Option<Binary>,
            param: u32,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

pub mod mismatched_param_arity {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=success)]
        fn first_reply(
            &self,
            _ctx: ReplyCtx,
            _data: Option<Binary>,
            param: String,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=failure)]
        fn second_reply(
            &self,
            _ctx: ReplyCtx,
            _data: Option<Binary>,
            param: String,
            param: u32,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

fn main() {}
