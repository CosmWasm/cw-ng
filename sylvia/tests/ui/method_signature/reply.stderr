error: Mismatched parameter in reply handlers.

         = note: Parameters for the `on_instantiated` handler have to be the same.
         = note: Previous parameter defined for the `on_instantiated` handler.

  --> tests/ui/method_signature/reply.rs:26:13
   |
26 |             param: String,
   |             ^^^^^

error: Mismatched quantity of method parameters.

         = note: Both `on_instantiated` handlers should have the same number of parameters.
         = note: Previous definition of on_instantiated handler.

  --> tests/ui/method_signature/reply.rs:60:12
   |
60 |         fn first_reply(
   |            ^^^^^^^^^^^
